rllib/offline/estimators/fqe_torch_model.py

```diff
@@ -130,33 +130,32 @@ def train_q(self, batch: SampleBatch) -> TensorType:
                 )
                 dones = torch.tensor(minibatch[SampleBatch.DONES], device=self.device)
 
-                # Neccessary if policy uses recurrent/attention model
-                num_state_inputs = 0
-                for k in batch.keys():
-                    if k.startswith("state_in_"):
-                        num_state_inputs += 1
-                state_keys = ["state_in_{}".format(i) for i in range(num_state_inputs)]
-
-                # Compute action_probs for next_obs as in FQE
-                all_actions = torch.zeros([minibatch.count, self.policy.action_space.n])
-                all_actions[:] = torch.arange(self.policy.action_space.n)
-                next_action_prob = self.policy.compute_log_likelihoods(
-                    actions=all_actions.T,
-                    obs_batch=next_obs,
-                    state_batches=[minibatch[k] for k in state_keys],
-                    prev_action_batch=minibatch[SampleBatch.ACTIONS],
-                    prev_reward_batch=minibatch[SampleBatch.REWARDS],
-                    actions_normalized=False,
-                )
-                next_action_prob = (
-                    torch.exp(next_action_prob.T).to(self.device).detach()
-                )
-
+                # Compute Q-values for current obs
                 q_values, _ = self.q_model({"obs": obs}, [], None)
                 q_acts = torch.gather(q_values, -1, actions.unsqueeze(-1)).squeeze(-1)
+
+                # Compute pi(a | s) for each action a in policy.action_space
+                next_action_probs = []
+                tmp_minibatch = minibatch.copy()
+                for i in range(self.policy.action_space.n):
```

Comment:
**todo (blocking)**: I feel like this can be written more elegantly. Here is one suggestion: You need to work with the entire action distribution here not the probability mass of a single action. So if you call distribution_fn() of the policy you are expected to get a categorical distribution. So you have access to the probability mass there and can re-use it with one single forward pass. Right now you are doing a forward pass basically for every possible action in the action space.

Reply:
Done